Add Menu and Dropdown design guidelines

mcarrano opened the pull request:
This PR adds new design guidelines documents for the Dropdown and Menu components. It also includes updates to the Select design guidelines, adding an Elements and variations section before usage to make this more similar to the Dropdown guidelines page.

Closes #2115 
Closes #2118

patternfly-build commented:
PF3 preview: https://patternfly-org-pr-2241-v3.surge.sh/v3

mmenestr commented on packages/v4/src/content/design-guidelines/components/dropdown/dropdown.md:
```diff
@@ -2,4 +2,39 @@
 ---
-Use a **dropdown** when you want to present a list of actions in a limited space.
+
+A **dropdown** component is used to present a menu of actions or links in a constrained space that will trigger a process or navigate to a new location. See Select component [link] if you want to make one or more selections from a list of items in a value list.
```
Need to edit the "select component [link]" to:
"See [Select component](/components/select/) if you want..."

mmenestr replied:
Also change beginning to "A **Dropdown** is used to..." - so capitalize dropdown, and remove the word "component". This is just so it's more in line with the start of other guidelines!

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+section: components
+---
+
+A menu component that can be used in a variety of contexts whenever the user needs to choose between multiple values, options, or actions. A menu can be opened in a [dropdown](/components/dropdown), a [select list](/components/select), or it can be revealed upon right clicking on a region within a page.
```
Change this too... "A **Menu** can be used in a variety...."
It doesn't really tell you what a menu is though, and more so how you can use it. Should we maybe add a sentence like "A **Menu** is a list of items inside a contained that can be used..." or something along those lines

mcarrano replied:
OK. I added an additional sentence. See what you think.

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+### Action menu
+An actions menu is used for presenting a list of actions or links.
+
+<img src="./img/basic-menu.png" alt="actions dropdown" width="863"/>
```
I feel like the pink headers in this image should be in sentence case? I'm guessing you title cased it because it's referring to the bullets, but I think it looks odd since they're just general "statements"

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+### Option select menus
+Use an option select menu when you want to persist the selected items. This is typically the case for select menus or settings menus that allow users to select between multiple options. See the [Select](/components/select) and [Options menu](/components/options-menu) components for more details. You may present a single or multiple groups of options within the same menu.
+
+<img src="./img/option-select-menu.png" alt="option select menus" width="863"/>
```
Are the checkboxes in these images meant to be grey? Or should they be blue? I know they are grey in the symbol library but I feel that this may be a bug. Been meaning to bring this up to Marie!

mcarrano replied:
Yeah, in the component examples they are blue. I can fix these, but it should also be fixed in the Sketch library.

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+4. **Descriptions (optional):** Descriptive text may be added below each menu item. This should only be used when the menu item label by itself may not be clear to all users. Keep descriptive text to two lines or less.
+
+### Option select menus
+Use an option select menu when you want to persist the selected items. This is typically the case for select menus or settings menus that allow users to select between multiple options. See the [Select](/components/select) and [Options menu](/components/options-menu) components for more details. You may present a single or multiple groups of options within the same menu.
```
I think it would help to link "groups of options" to that section in the doc --> [groups of options](#grouped-menus)

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+When filtering is being used, a search input will be added to the top of the menu and the list of items will be filtered as the user types.
+
+### Favoriting and other actions
+Favoriting or any other optional action can be associated with any menu item. In this case, making an item a favorite will duplicate in at the top of the menu. This may be useful when you have a long list of possible actions and want to “pin” a subset of items to the top of the menu.
```
little typo in here - should say "will duplicate **it**" not "duplicate **in**"

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+
+| **When you want to...** |                     |
+| ------------------- | ------------------- |
+| Expose a list of actions in a dropdown menu | Use an actions menu with actions and/or links. Add icons in front of menu items only if they will help to make items more recognizable. Include descriptive text for menu items that may require further definition. |
```
would take out the ellipses at the end of each use case, the first row doesn't have them, and some are alone on the next line which looks a little weird

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+| Expose a list of actions in a dropdown menu | Use an actions menu with actions and/or links. Add icons in front of menu items only if they will help to make items more recognizable. Include descriptive text for menu items that may require further definition. |
+| Create logical groupings of menu items... | Use separators to group items. Include group headings if you don’t feel like the meaning of the groups will be obvious. |
+| Expose a list of nested actions... | Use a fly-out menu. Any optional attributes of basic menus, including icons, groups, and descriptions can also be applied to flyout menus.|
+| Select from a list of options in a dropdown menu ... | Use an options select menu for selecting between values rather than actions. Icons, descriptions, and groupings can be used with option select menus to improve recognition or create logical option groups. |
```
This last case is a little confusing, are you talking about select lists, or drop downs? It seems like you are talking about select lists here so I would probably remove any mention of "drop downs" to prevent confusion!

Would rephrase to "select from a list of options in a select menu..." "Use an options select menu etc"

mcarrano replied:
This one I'm not sure about. I was using the word "dropdown" to refer to the type of menu generically rather than the component named "Dropdown." Options select menus are used in the Options menu component as well as the Select component. Ultimately we are looking to use the Menu component as part of a composable dropdown menu approach where you could put together custom dropdown menus. What if I just get rid of the work "dropdown" here and just say, "Select from a list of options?"

mmenestr replied:
Yes, I think that would be a good approach. I just want to steer clear of using the word "dropdown" unless it's referring to the actual dropdown component

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+| Select from a list of options in a dropdown menu ... | Use an options select menu for selecting between values rather than actions. Icons, descriptions, and groupings can be used with option select menus to improve recognition or create logical option groups. |
+
+## Content guidelines
+It creating menu item labels, keep in mind the following guidelines:
```
Typo - assuming you meant "in creating" not "it creating"

mmenestr commented on packages/v4/src/content/design-guidelines/components/select/select.md:
```diff
@@ -9,7 +9,29 @@ related: [
+
+2. **Menu:** The menu contains a list of options or values. Menus should optimally contain seven (7) or less items for easy scanning by the user. For longer menus, items may be grouped. See the [Menu component guidelines](/components/menu/design-guidelines) for more information about supported options, including use of icons, grouping, and descriptions. The Dropdown component includes many common ...
+
+### Select variations
```
From here to the usage header repeats a lot of what's already been added to the guidelines. I would definitely just merge them somehow

mmenestr replied:
Or maybe delete this who select variations - it's a little confusing too because typeahead and checkbox do very different things ! If anything i'd expected a checkbox vs multiple select comparison?

mcarrano replied:
I'm going to just delete this section. I added it at the last minute, but I see where it is redundant. I mainly wanted to just point out the basic elements and reference the menu guidelines as I do in the first section.

mmenestr commented on packages/v4/src/content/design-guidelines/components/select/select.md:
```diff
@@ -9,7 +9,29 @@ related: [
+
+1. **Toggle:** The toggle is used to open and close the menu. It takes on all of the behaviors of a button. The toggle label always reflects the currently selected item or may hold a placeholder if nothing is selected.
+
+2. **Menu:** The menu contains a list of options or values. Menus should optimally contain seven (7) or less items for easy scanning by the user. For longer menus, items may be grouped. See the [Menu component guidelines](/components/menu/design-guidelines) for more information about supported options, including use of icons, grouping, and descriptions. The Dropdown component includes many common ...
```
This is confusing - did you mean to change the second half of the text but forgot to? It seems like it's talking about drop downs even though this is the select guidelines?
I also would want to add a section on scrolling here, or some other recommendations for when the select list is too long. In most case, if someones list is too long, grouping it wouldn't help with that issue! I'd expect a recommendation of making the menu scrollable, or adding a "see more" button at the bottom to view more options. I know this is a common battle/question in c.rh.c

mcarrano replied:
Yes, this is confusing. I did a quick copy and paste without reading this thoroughly. Hopefully it's better now.

mcarrano commented:
@mmenestr This is building now. I addressed your comments. Please re-review.

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+
+1. **Group heading (optional):** Add a group heading when you want to name the group.
+
+2. **Separator:** Separators are horizontal dividers that help to group the menu items by clearly showing where one group ends and the next begins.
```
@mcarrano Do you disagree with this^?

mcarrano replied:
I can change this to optional @mmenestr . I guess it raises a question of when to use a separator and when not to. I looked at the examples for Select. I think the grouped checkbox select examples are fine without separators, but for other examples it would be hard to see the grouping without them. I think this is partly because the style of the section headers is not that distinct from other menu items. Thoughts?

mmenestr replied:
That's a good point, maybe we say that in dropdown menus it makes sense to use a separator where as in select lists it's not necessary because there is more visual distinction!

mmenestr commented on packages/v4/src/content/design-guidelines/components/select/select.md:
```diff
@@ -9,7 +9,17 @@ related: [
+
+A select component consists of a toggle control to open and close a menu of actions or links. Selects differ from [dropdowns](/components/dropdown) in that they persist selection where dropdowns are typically used to present a list of actions or links.
+
+<img src="./img/basic-select.png" alt="basic select menu" width="303"/>
```
Need to change this image too to use a blue checkmark instead of grey!

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+### Option select menus
+Use an option select menu when you want to persist the selected items. This is typically the case for select menus or settings menus that allow users to select between multiple options. See the [Select](/components/select) and [Options menu](/components/options-menu) components for more details. You may present a single or multiple [groups of options](#grouped-menus) within the same menu.
+
+<img src="./img/option-select-menu.png" alt="option select menus" width="863"/>
```
I think the checkmarks in this image are the wrong blue - Marie updated the symbol now so you should be able to use that now!

mcarrano replied:
Wow. You've got a good eye! Fixed now.

mmenestr commented:
Revisions looked good! Just left a few more comments for things I noticed

mcarrano commented:
@mmenestr I made a few more updates. Can you review one more time? Thanks.

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+### Grouped menus
+You may decide to group menu items to associate related items and/or to indicate a hierarchy of items. Items may be grouped using group headings, separators, or both.
+
+<img src="./img/grouped-menu.png" alt="grouped menu" width="348"/>
```
This image looks tiny - I think you need to update the width to match the new artboard!

mmenestr commented on packages/v4/src/content/design-guidelines/components/menu/menu.md:
```diff
@@ -0,0 +1,82 @@
+
+1. **Group heading (optional):** Add a group heading when you want to name the group.
+
+2. **Separator (optional):** Separators are horizontal dividers that help to group the menu items by clearly showing where one group ends and the next begins. You should consider whether a separator is needed to create visually distinct groups. In the first example above, a separator is used to create separation between the two groups of items. In the second example, The separator is not needed b ...
```
T is un=necessarily capitalize in the last sentence, where it says "The seperator" after the comma

mmenestr commented on packages/v4/src/content/design-guidelines/components/select/select.md:
```diff
@@ -19,22 +29,22 @@ PatternFly offers several variations of select lists:
 * [Typeahead](#typeahead)
 
-#### Single select 
+#### Single select
```
Realizing now that these headings should probably be made H3s instead of H4s, because right now the table of contents looks very odd

mcarrano replied:
Can we open a new issue for that and get this one merged?

mcarrano commented:
@mmenestr Here it is one more time. Made fixes to the Menu guidelines to address the two issues above.

mmenestr commented:
@mcarrano the first image in select list still has the grey checkmark - after that should be ready to go! Sorry I didn't notice it yesterday afternoon

mcarrano commented:
@mmenestr I updated that last image on the Select page and also fixed the heading styles as you suggested while I was there. Please review one more time. Thanks.

patternfly-build commented:
Your changes have been released in: 
 - patternfly-org-4@4.2.68

Thanks for your contribution! :tada:
